### PR TITLE
Apply transforms to raw data when MNI6 derivatives aren't available

### LINKED ISSUE #11: Add `mem_gb` to SUSAN and MELODIC nodes to avoid memory errors
I have been having some issues with `BrokenProcessPool` errors (i.e., out-of-memory errors), and these usually occur around the `melodic` and `susan` commands. Would it be possible to add a `mem_gb` of a few gigabytes to both of these, so that the `nipype` scheduler knows not to start too many at the same time?

https://github.com/nipreps/fmriprep/blob/a45ea486a8aa58541add3370a60ae87a70af8efb/fmriprep/workflows/bold/confounds.py#L668-L673

### LINKED ISSUE #57: Support raw dataset + derivatives with transforms to MNI152NLin6Asym
This stems from #42, where I noted that the following combinations of raw and derivative datasets failed to work:

- [ ]  nomni6-func-raw
- [ ]  nomni6-func-noraw
- [ ]  nomni6-nofunc-raw
- [ ]  nomni6-nofunc-noraw (don't expect to work)

### PR DESCRIPTION
**Pull request by tsalo**
Closes #57. I tested this locally against running with MNI6 derivatives, and the results are different. I am a little worried that I didn't implement the transforms properly or something.

To do:

- [x] Ensure boilerplate is accurate
- [ ] Load STC slice ref from preprocessed BOLD data (need to find preproc BOLD data in _any_ space). Moved to #60.

## Changes proposed in this pull request

- Add `ignore` to the Config object so users can actually ignore stuff.
- Ignore fieldmaps by default, since I don't know how to grab them yet. Will revert this later.
- Add mem_gb to SUSAN and MELODIC nodes (closes #11).
- Allow processing of raw BOLD data when MNI6 derivatives aren't available.
    - We still need an anat-to-MNI6 transform.
    - This runs STC, then `init_bold_volumetric_resample_wf`, using raw BOLD and derivatives I think should be available at least in `full` mode with `--cifti-output`.

### CODE REVIEW

File: src/fmripost_aroma/interfaces/confounds.py [hunk 1]
```diff
@@ -101,9 +101,7 @@ def _get_ica_confounds(mixing, aroma_features, skip_vols, newpath=None):
 
     signal_mixing_arr = mixing_arr[signal_ics, :].T
     orthaggr_mixing_arr = aggr_mixing_arr.copy()
-    orthaggr_mixing_arr = aggr_mixing_arr - np.dot(
-        np.dot(np.linalg.pinv(good_ic_arr), good_ic_arr), aggr_mixing_arr
-    )
+
```
Comment (tsalo) [inline src/fmripost_aroma/interfaces/confounds.py:104]: Bug. Move to a separate PR.

File: src/fmripost_aroma/workflows/base.py [hunk 2]
```diff
@@ -363,25 +373,96 @@ def init_single_run_wf(bold_file):
         ),
         name='mni6_buffer',
     )
-    if 'bold_mni152nlin6asym' not in functional_cache:
+
+    if ('bold_mni152nlin6asym' not in functional_cache) and ('bold_raw' in functional_cache):
         # Resample to MNI152NLin6Asym:res-2, for ICA-AROMA classification
-        run_stc = (
-            bool(bold_metadata.get('SliceTiming')) and 'slicetiming' not in config.workflow.ignore
```
Comment (tsalo) [inline src/fmripost_aroma/workflows/base.py:369]: Add this back in.

File: src/fmripost_aroma/workflows/base.py [hunk 3]
```diff
@@ -363,25 +373,96 @@ def init_single_run_wf(bold_file):
...
+        )
+        run_stc = 'slicetiming' not in config.workflow.ignore
+        if run_stc:
+            bold_stc_wf = init_bold_stc_wf(
+                mem_gb=mem_gb,
+                metadata=bold_metadata,
+                name='resample_stc_wf',
+            )
+            # TODO: Grab skip-vols from confounds file
+            bold_stc_wf.inputs.inputnode.skip_vols = config.workflow.dummy_scans or 0
```
Comment (tsalo) [inline src/fmripost_aroma/workflows/base.py:402]: Should I grab skip-vols from the confounds file?

File: src/fmripost_aroma/workflows/base.py [hunk 4]
```diff
@@ -363,25 +373,96 @@ def init_single_run_wf(bold_file):
...
-                ('outputnode.bold_std', 'inputnode.bold_mni152nlin6asym'),
-                ('outputnode.bold_mask_std', 'inputnode.bold_mask_mni152nlin6asym'),
-            ]),
+            # Resample BOLD to MNI152NLin6Asym, may duplicate bold_std_wf above
+            # XXX: Ignoring the field map for now
+            # (inputnode, bold_MNI6_wf, [
+            #     ('fmap_ref', 'inputnode.fmap_ref'),
+            #     ('fmap_coeff', 'inputnode.fmap_coeff'),
+            #     ('fmap_id', 'inputnode.fmap_id'),
+            # ]),
```
Comment (tsalo) [inline src/fmripost_aroma/workflows/base.py:446]: Need to figure out how to grab the field map

File: src/fmripost_aroma/workflows/resampling.py [hunk 5]
```diff
@@ -16,7 +16,7 @@ def init_resample_volumetric_wf(
     Parameters
```
Comment (tsalo) [inline src/fmripost_aroma/workflows/resampling.py:16]: I can just remove this workflow for now.

File: src/fmripost_aroma/interfaces/confounds.py (hunk 1 above)
Comment (tsalo) [reply src/fmripost_aroma/interfaces/confounds.py:104]: Moved to #59.

File: src/fmripost_aroma/workflows/base.py (hunk 2 above)
Comment (tsalo) [reply src/fmripost_aroma/workflows/base.py:369]: Done in [2f1dc8b](https://github.com/nipreps/fmripost-aroma/pull/58/commits/2f1dc8ba56d80397c297bde25d6e94bb7f73b4a7)

**Comment by tsalo** [comment]
# With MNI6

image

# Without MNI6 (i.e., warping raw)

That first component really throws me.

image

**Comment by tsalo** [comment]
I realized I wasn't passing along the STCed BOLD data to the MNI6 workflow, so I fixed that, but it didn't bring the results in line with the MNI6-derivatives outputs.

**Comment by tsalo** [comment]
I finally took a look at the warped data from my PR and the data are in the wrong space. The MNI-space BOLD from fMRIPrep is in red and overlaps well with the brain mask. The warped BOLD from this PR is in blue and does not overlap with the brain mask.

image

**Comment by tsalo** [comment]
The transforms being used for the `resample` (`ResampleSeries`) step of `bold_MNI6_wf` (`init_bold_volumetric_resample_wf`) seem right to me:

1. sub-01/ses-01/func/sub-01_ses-01_task-rest_from-orig_to-boldref_mode-image_desc-hmc_xfm.txt
2. sub-01/ses-01/func/sub-01_ses-01_task-rest_from-boldref_to-T1w_mode-image_desc-coreg_xfm.txt
3. sub-01/ses-01/anat/sub-01_ses-01_from-T1w_to-MNI152NLin6Asym_mode-image_xfm.h5

Comparing to what's in the fMRIPrep working directory:

1. sub-01/ses-01/func/sub-01_ses-01_task-rest_from-orig_to-boldref_mode-image_desc-hmc_xfm.txt
2. sub-01/ses-01/func/sub-01_ses-01_task-rest_from-boldref_to-T1w_mode-image_desc-coreg_xfm.txt
3. sub-01/ses-01/anat/sub-01_ses-01_from-T1w_to-MNI152NLin6Asym_mode-image_xfm.h5

**Comment by tsalo** [comment]
Okay, I think it's an issue in the Docker image. I ran ResampleSeries locally and the results look good for boldref-to-T1w-to-MNI6 _and_ orig-boldref-to-T1w-to-MNI6. I then tried it on the same files, with the same commands, in the Docker image, and, while boldref-to-T1w-to-MNI6 looks good, orig-boldref-to-T1w-to-MNI6 does not. So basically the HMC part is not operating correctly within the Docker image.

I assume the problem is the Python version in my Docker image.

**Comment by tsalo** [comment]
@effigies I bumped my local nitransforms from 23.0.1 to 24.0.0 and the orig-to-boldref-to-T1w-to-MNI6 result started looking bad. Is there some chance an issue with using fMRIPrep's HMC transforms was introduced in nitransforms 24.0.0?

**Comment by effigies** [comment]
Can you decompose that? I don't think anything should have changed in HMC, but we did fix the parsing of the ITK warps, which might have broken our workaround in nipreps...

**Comment by effigies** [comment]
Okay. Damn. We probably need to roll back nitransforms in the fmriprep 24.1.0 branch.

**Comment by effigies** [comment]
Actually, would you mind re-testing with fmriprep 24.1.0rc0, to verify that the problem exhibits there as well?

**Comment by tsalo** [comment]
I can run the subject through the full pipeline later today, but in case it's informative, I pulled nipreps/fmriprep:24.0.0 and nipreps/fmriprep:24.1.0rc0 and ran just the ResampleSeries command with those. The results look good for 24.0.0, but not for 24.1.0rc0.

**Comment by tsalo** [comment]
I've got fMRIPrep running with 24.1.0rc0 and I'll update this thread with my results once it finishes.

The good news is that I pinned `nitransforms` to 23.0.1 and rebuilt the fMRIPost-AROMA container, and the results look solid to me, so I'm going to merge this PR.

**Comment by tsalo** [comment]
We're having server issues right now so it might be a few days before I can resubmit my fMRIPrep 24.1.0rc0 job.
